File: Excel/900-999/993/Challenge 993.R
library(tidyverse)
library(readxl)

path <- "900-999/993/993 Pivot.xlsx"
input <- read_excel(path, range = "A1:A21")
test <- read_excel(path, range = "C2:D8")

priority_w <- c(H = .5, M = .3, L = .1)
channel_w <- c(E = .4, P = .2, C = .1)
currency_fx <- c(USD = 94, EUR = 130, INR = 1)

result <- input %>%
  mutate(
    City = str_match(Data, "LOC\\(([^)]+)\\)")[, 2],
    Currency = str_match(Data, "TX\\{[^|]+\\|([^|]+)\\|")[, 2],
    Amount = as.numeric(str_match(Data, "TX\\{[^|]+\\|[^|]+\\|([^|]+)\\|")[,
      2
    ]),
    Priority = str_match(Data, "P:([^/]+)")[, 2],
    Channel = str_match(Data, "C:([^/]+)")[, 2],
    Weighted = Amount *
      currency_fx[Currency] *
      (coalesce(priority_w[Priority], 0) + coalesce(channel_w[Channel], 0))
  ) %>%
  summarise(
    `Weighted Total (INR)` = sum(Weighted),
    .by = City
  ) %>%
  arrange(City)

result == test
# One value is different.

File: Excel/900-999/993/Challenge 993.py
import pandas as pd

path = "900-999/993/993 Pivot.xlsx"
input = pd.read_excel(path, usecols="A", nrows=21)
test = pd.read_excel(path, usecols="C:D", nrows=6, skiprows=1)

priority_w = {"H": .5, "M": .3, "L": .1}
channel_w  = {"E": .4, "P": .2, "C": .1}
currency_fx = {"USD": 94, "EUR": 130, "INR": 1}

out = (
    input["Data"]
    .str.extract(
        r"LOC\((?P<City>[^)]+)\)-TX\{[^|]+\|(?P<Currency>[A-Z]{3})\|(?P<Amount>\d+)\|[^}]+\}"
    )
    .join(input["Data"].str.extract(r"\bP:(?P<Priority>[HML])\b"))
    .join(input["Data"].str.extract(r"\bC:(?P<Channel>[EPC])\b"))
    .assign(
        Amount=lambda x: pd.to_numeric(x["Amount"], errors="coerce").fillna(0).astype("int64"),
        Weighted=lambda x: (
            x["Amount"]
            * x["Currency"].map(currency_fx).fillna(0)
            * (
                x["Priority"].map(priority_w).fillna(0)
                + x["Channel"].map(channel_w).fillna(0)
            )
        ).round().astype("int64")
    )
    .groupby("City", as_index=False)["Weighted"]
    .sum()
    .rename(columns={"Weighted": "Weighted Total (INR)"})
    .sort_values("City")
)

print(out == test)
# One Value is different. 